modules/45_carbonprice/functionalForm/postsolve.gms

```diff
@@ -6,92 +6,233 @@
+    );
+    display p45_taxCO2eq_anchor, p45_taxCO2eq_anchor_until2150, p45_factorRescale_taxCO2, p45_factorRescale_taxCO2_Funneled;
+
+    !! Apply CO2 tax rescale factor
+    p45_taxCO2eq_anchor_until2150(t) = max(1* sm_DptCO2_2_TDpGtC, p45_taxCO2eq_anchor_until2150(t) * p45_factorRescale_taxCO2_Funneled(iteration) );
+    !! Save difference between iterations for debugging (note that p45_taxCO2eq_anchor was not yet rescaled up to this point)
+    pm_taxCO2eq_anchor_iterationdiff(t) = max(1* sm_DptCO2_2_TDpGtC, p45_taxCO2eq_anchor(t) * p45_factorRescale_taxCO2_Funneled(iteration) ) - p45_taxCO2eq_anchor(t);
+    o45_taxCO2eq_anchor_iterDiff_Itr(iteration) = pm_taxCO2eq_anchor_iterationdiff("2100");
+    display p45_taxCO2eq_anchor_until2150, pm_taxCO2eq_anchor_iterationdiff, o45_taxCO2eq_anchor_iterDiff_Itr;
+
+    !! If functionalForm is linear, re-adjust global anchor trajectory to go through the point (cm_taxCO2_historicalYr, cm_taxCO2_historical) 
+$ifThen.taxCO2functionalForm4 "%cm_taxCO2_functionalForm%" == "linear"
+    p45_taxCO2eq_anchor_until2150(t) = s45_taxCO2_historical 
+        + (sum(t2$(t2.val eq cm_peakBudgYr), p45_taxCO2eq_anchor_until2150(t2)) - s45_taxCO2_historical) / (cm_peakBudgYr - s45_taxCO2_historicalYr) !! Yearly increase of CO2 price that interpolates between cm_taxCO2_historical in cm_taxCO2_historicalYr and p45_taxCO2eq_anchor_until2150 in peak year
                                       * (t.val - s45_taxCO2_historicalYr) ;
-p45_taxCO2eq_anchor(t)$(t.val gt 2110) = p45_taxCO2eq_anchor("2110");
-$endIf.taxCO2functionalForm2
-	);
-);
+    display p45_taxCO2eq_anchor_until2150;
+$endIf.taxCO2functionalForm4 
+
+    !! Use rescaled p45_taxCO2eq_anchor_until2150 as starting point for re-defining p45_taxCO2eq_anchor
+    p45_taxCO2eq_anchor(t) = p45_taxCO2eq_anchor_until2150(t);
+    
+    if(cm_iterative_target_adj = 9, !! After cm_peakBudgYr, the global anchor trajectory increases linearly with fixed annual increase given by cm_taxCO2_IncAfterPeakBudgYr
+      p45_taxCO2eq_anchor(t)$(t.val gt cm_peakBudgYr) = sum(t2$(t2.val eq cm_peakBudgYr), p45_taxCO2eq_anchor_until2150(t2)) !! CO2 tax in peak budget year
+                                                  + (t.val - cm_peakBudgYr) * cm_taxCO2_IncAfterPeakBudgYr * sm_DptCO2_2_TDpGtC;  !! increase by cm_taxCO2inc_after_peakBudgYr per year 
+    );  
+    !! Always set carbon price constant after 2100 to prevent huge taxes after 2100 and the resulting convergence problems
+    p45_taxCO2eq_anchor(t)$(t.val gt 2100) = p45_taxCO2eq_anchor("2100");
+    display p45_taxCO2eq_anchor;
+
+  ); !! if( (o_modelstat ne 2) OR (abs(cm_budgetCO2from2020 - s45_actualbudgetco2) < 2) OR (ord(iteration) = cm_iteration_max), 
+); !! if((cm_iterative_target_adj eq 5) OR (cm_iterative_target_adj eq 9),
+
+
+*** -------B: checking the peak timing, if cm_peakBudgYr is still correct or needs to be shifted-----------------------
```

Comment:
This part is about peak year adjustment, thus only used with cm_iterative_target_adj = 9.